NDTensors/src/tensor/tensor.jl

```diff
@@ -347,7 +353,10 @@ end
 # Some generic getindex and setindex! functionality
 #
 
-@propagate_inbounds @inline setindex!!(T::Tensor, x, I...) = setindex!(T, x, I...)
+@propagate_inbounds @inline function setindex!!(T::Tensor, x, I...)
+  T = allocate(T, typeof(x))
```

Comment:
This relates to the message I sent on slack. What happens if the element type of `T` is already specified (say constructed with `ITensor(Float32, i, j)`)? I think we should preserve the element type if it was already specified, and only set it to `typeof(x)` if it is `UnspecifiedZero`, otherwise it could lead to surprising behavior where a real tensor could get switched to complex, for example.

Maybe it is best to leave that logic out of `allocate`, and instead use the semantics:
```julia
T = allocate(set_eltype_if_unspecified(T, typeof(x)))
```